File: server/internal/ddb/schema.go
package ddb

import (
	"log"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/service/dynamodb"
	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbiface"
)

const (
	tablePrimaryKey        = "PK"
	tableSortKey           = "SK"
	gsi1Name               = "GSI1"
	gsi1PrimaryKey         = "GSI1PK"
	gsi1SortKey            = "GSI1SK"
	unsettledTxnsIndexName = "UnsettledTransactions"
	unsettledTxnsIndexPK   = "PK"
	unsettledTxnsIndexSK   = "Unsettled"
)

// CreateTable creates a table of the expenseus schema with the given name.
func CreateTable(d dynamodbiface.DynamoDBAPI, name string) error {
	_, err := d.CreateTable(&dynamodb.CreateTableInput{
		AttributeDefinitions: []*dynamodb.AttributeDefinition{
			{
				AttributeName: aws.String(tablePrimaryKey),
				AttributeType: aws.String("S"),
			},
			{
				AttributeName: aws.String(tableSortKey),
				AttributeType: aws.String("S"),
			},
			{
				AttributeName: aws.String(gsi1PrimaryKey),
				AttributeType: aws.String("S"),
			},
			{
				AttributeName: aws.String(gsi1SortKey),
				AttributeType: aws.String("S"),
			},
			{
				AttributeName: aws.String(unsettledTxnsIndexSK),
				AttributeType: aws.String("S"),
			},
		},
		KeySchema: []*dynamodb.KeySchemaElement{
			{
				AttributeName: aws.String(tablePrimaryKey),
				KeyType:       aws.String("HASH"),
			},
			{
				AttributeName: aws.String(tableSortKey),
				KeyType:       aws.String("RANGE"),
			},
		},
		GlobalSecondaryIndexes: []*dynamodb.GlobalSecondaryIndex{
			{
				IndexName: aws.String(gsi1Name),
				KeySchema: []*dynamodb.KeySchemaElement{
					{
						AttributeName: aws.String(gsi1PrimaryKey),
						KeyType:       aws.String("HASH"),
					},
					{
						AttributeName: aws.String(gsi1SortKey),
						KeyType:       aws.String("RANGE"),
					},
				},
				Projection: &dynamodb.Projection{ProjectionType: aws.String(dynamodb.ProjectionTypeAll)},
				ProvisionedThroughput: &dynamodb.ProvisionedThroughput{
					ReadCapacityUnits:  aws.Int64(1),
					WriteCapacityUnits: aws.Int64(1),
				},
			},
			{
				IndexName: aws.String(unsettledTxnsIndexName),
				KeySchema: []*dynamodb.KeySchemaElement{
					{
						AttributeName: aws.String(unsettledTxnsIndexPK),
						KeyType:       aws.String("HASH"),
					},
					{
						AttributeName: aws.String(unsettledTxnsIndexSK),
						KeyType:       aws.String("RANGE"),
					},
				},
				Projection: &dynamodb.Projection{ProjectionType: aws.String(dynamodb.ProjectionTypeAll)},
				ProvisionedThroughput: &dynamodb.ProvisionedThroughput{
					ReadCapacityUnits:  aws.Int64(1),
					WriteCapacityUnits: aws.Int64(1),
				},
			},
		},
		ProvisionedThroughput: &dynamodb.ProvisionedThroughput{
			ReadCapacityUnits:  aws.Int64(1),
			WriteCapacityUnits: aws.Int64(1),
		},
		TableName: aws.String(name),
	})
	if err != nil {
		return err
	}

	log.Println("successfully created the table", name)
	return nil
}

File: client/ddb/schema.ts
import {
  CreateTableCommand,
  DeleteTableCommand,
  DynamoDBClient,
  ProjectionType,
  ResourceInUseException,
} from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient } from '@aws-sdk/lib-dynamodb';

// TODO: load from env variable
const endpoint = process.env.DDB_ENDPOINT;
const accessKeyId = process.env.DDB_ACCESS_KEY_ID ?? '';
const secretAccessKey = process.env.DDB_SECRET_ACCESS_KEY ?? '';
export const tableName = process.env.DDB_TABLE ?? 'expenseus-default';

export type DDBWithConfig = {
  ddb: DynamoDBDocumentClient;
  tableName: string;
};

export function setUpDdb(tableName: string) {
  const ddbClient = new DynamoDBClient({
    endpoint,
    credentials: {
      accessKeyId,
      secretAccessKey,
    },
  });
  const ddbDocClient = DynamoDBDocumentClient.from(ddbClient);
  return {
    ddb: ddbDocClient,
    tableName,
  };
}

export const tablePartitionKey = 'PK',
  tableSortKey = 'SK',
  gsi1Name = 'GSI1',
  gsi1PartitionKey = 'GSI1PK',
  gsi1SortKey = 'GSI1SK',
  unsettledTxnsIndexName = 'UnsettledTransactions',
  unsettledTxnsIndexPK = 'PK',
  unsettledTxnsIndexSK = 'Unsettled';

function createTableCommand(tableName: string) {
  return new CreateTableCommand({
    TableName: tableName,
    KeySchema: [
      {
        AttributeName: tablePartitionKey,
        KeyType: 'HASH',
      },
      {
        AttributeName: tableSortKey,
        KeyType: 'RANGE',
      },
    ],
    AttributeDefinitions: [
      {
        AttributeName: tablePartitionKey,
        AttributeType: 'S',
      },
      {
        AttributeName: tableSortKey,
        AttributeType: 'S',
      },
      {
        AttributeName: gsi1PartitionKey,
        AttributeType: 'S',
      },
      {
        AttributeName: gsi1SortKey,
        AttributeType: 'S',
      },
      {
        AttributeName: unsettledTxnsIndexSK,
        AttributeType: 'S',
      },
    ],
    GlobalSecondaryIndexes: [
      {
        IndexName: gsi1Name,
        KeySchema: [
          {
            AttributeName: gsi1PartitionKey,
            KeyType: 'HASH',
          },
          {
            AttributeName: gsi1SortKey,
            KeyType: 'RANGE',
          },
        ],
        Projection: {
          ProjectionType: ProjectionType.ALL,
        },
        ProvisionedThroughput: {
          ReadCapacityUnits: 1,
          WriteCapacityUnits: 1,
        },
      },
      {
        IndexName: unsettledTxnsIndexName,
        KeySchema: [
          {
            AttributeName: unsettledTxnsIndexPK,
            KeyType: 'HASH',
          },
          {
            AttributeName: unsettledTxnsIndexSK,
            KeyType: 'RANGE',
          },
        ],
        Projection: {
          ProjectionType: ProjectionType.ALL,
        },
        ProvisionedThroughput: {
          ReadCapacityUnits: 1,
          WriteCapacityUnits: 1,
        },
      },
    ],
    ProvisionedThroughput: {
      ReadCapacityUnits: 1,
      WriteCapacityUnits: 1,
    },
  });
}

export async function createTableIfNotExists(tableName: string) {
  const d = setUpDdb(tableName);
  try {
    await d.ddb.send(createTableCommand(tableName));
  } catch (err) {
    if (err instanceof ResourceInUseException) {
      console.log(`Table exists. Continuing...`);
      return;
    }
    throw err;
  }
}

export async function TESTONLY_deleteTable(tableName: string) {
  const d = setUpDdb(tableName);
  try {
    await d.ddb.send(
      new DeleteTableCommand({
        TableName: tableName,
      }),
    );
  } catch (err) {
    if (err instanceof ResourceInUseException) {
      console.error('Tried to delete a non-existent table. Continuing...');
    }
    throw err;
  }
}
